internal/ls/utilities.go

```diff
@@ -0,0 +1,376 @@
+	if nextToken == nil || nextToken.Kind == ast.KindCloseBraceToken {
+		return true
+	}
+
+	startLine, _ := scanner.GetLineAndCharacterOfPosition(file, node.End())
+	endLine, _ := scanner.GetLineAndCharacterOfPosition(file, astnav.GetStartOfNode(nextToken, file, false /*includeJSDoc*/))
+	return startLine != endLine
+}
+
+func isNonContextualKeyword(token ast.Kind) bool {
+	return ast.IsKeywordKind(token) && !ast.IsContextualKeyword(token)
+}
+
+func probablyUsesSemicolons(file *ast.SourceFile) bool {
+	withSemicolon := 0
+	withoutSemicolon := 0
+	nStatementsToObserve := 5
+
+	var visit func(node *ast.Node) bool
+	visit = func(node *ast.Node) bool {
+		if syntaxRequiresTrailingSemicolonOrASI(node.Kind) {
+			lastToken := getLastToken(node, file)
+			if lastToken != nil && lastToken.Kind == ast.KindSemicolonToken {
+				withSemicolon++
+			} else {
+				withoutSemicolon++
+			}
+		} else if syntaxRequiresTrailingCommaOrSemicolonOrASI(node.Kind) {
+			lastToken := getLastToken(node, file)
+			if lastToken != nil && lastToken.Kind == ast.KindSemicolonToken {
+				withSemicolon++
+			} else if lastToken != nil && lastToken.Kind != ast.KindCommaToken {
+				lastTokenLine, _ := scanner.GetLineAndCharacterOfPosition(
+					file,
+					astnav.GetStartOfNode(lastToken, file, false /*includeJSDoc*/))
+				nextTokenLine, _ := scanner.GetLineAndCharacterOfPosition(
+					file,
+					scanner.GetRangeOfTokenAtPosition(file, lastToken.End()).Pos())
+				// Avoid counting missing semicolon in single-line objects:
+				// `function f(p: { x: string /*no semicolon here is insignificant*/ }) {`
+				if lastTokenLine != nextTokenLine {
+					withoutSemicolon++
+				}
+			}
+		}
+
+		if withSemicolon+withoutSemicolon >= nStatementsToObserve {
+			return true
+		}
+
+		return node.ForEachChild(visit)
+	}
+
+	file.ForEachChild(visit)
+
+	// One statement missing a semicolon isn't sufficient evidence to say the user
+	// doesn't want semicolons, because they may not even be done writing that statement.
+	if withSemicolon == 0 && withoutSemicolon <= 1 {
+		return true
+	}
+
+	// If even 2/5 places have a semicolon, the user probably wants semicolons
+	return withSemicolon/withoutSemicolon > 1/nStatementsToObserve
+}
+
+var typeKeywords []ast.Kind = []ast.Kind{
+	ast.KindAnyKeyword,
+	ast.KindAssertsKeyword,
+	ast.KindBigIntKeyword,
+	ast.KindBooleanKeyword,
+	ast.KindFalseKeyword,
+	ast.KindInferKeyword,
+	ast.KindKeyOfKeyword,
+	ast.KindNeverKeyword,
+	ast.KindNullKeyword,
+	ast.KindNumberKeyword,
+	ast.KindObjectKeyword,
+	ast.KindReadonlyKeyword,
+	ast.KindStringKeyword,
+	ast.KindSymbolKeyword,
+	ast.KindTypeOfKeyword,
+	ast.KindTrueKeyword,
+	ast.KindVoidKeyword,
+	ast.KindUndefinedKeyword,
+	ast.KindUniqueKeyword,
+	ast.KindUnknownKeyword,
+}
+
+func isTypeKeyword(kind ast.Kind) bool {
+	return slices.Contains(typeKeywords, kind)
+}
+
+// Returns a map of all names in the file to their positions.
+// !!! cache this
```

Comment:
If you eventually need to do this, you'll probably just want to shove this onto the SourceFile like lineMap.